command/eval_delete.go

```diff
@@ -0,0 +1,525 @@
+	}
+	return e.batchDelete([]*api.Evaluation{evalInfo})
+}
+
+// handleFlagFilterDelete handles deletion of evaluations discovered using
+// filter flags. It is unknown how many will match the operator criteria so
+// this function batches lookup and delete requests into sensible numbers.
+//
+// The function is recursive and will run until it has found and deleted all
+// evals which matched the criteria. It is possible for a batch of evals to be
+// deleted, but a subsequent call may fail due to temporary issues. The command
+// is idempotent and may be re-run safely if this occurs.
+func (e *EvalDeleteCommand) handleFlagFilterDelete(nextToken, filter string) (int, error) {
+
+	// Generate the query options using the passed next token and filter. The
+	// per page value is less than the total number we can include in a single
+	// delete request. This keeps the maximum size of the return object at a
+	// reasonable size.
+	//
+	// The JSON serialized evaluation API object is 350-380B in size.
+	// 2426 * 380B (3.8e-4 MB) = 0.92MB. We may want to make this configurable
+	// in the future, but this is counteracted by the CLI logic which will loop
+	// until the user tells it to exit, or all evals matching the filter are
+	// deleted. 2426 * 3 falls below the maximum limit for eval IDs in a single
+	// delete request (set by MaxEvalIDsPerDeleteRequest).
+	opts := &api.QueryOptions{
+		Filter:    filter,
+		PerPage:   2426,
+		NextToken: nextToken,
+	}
+
+	var evalsToDelete []*api.Evaluation
+
+	// Call List 3 times to accumulate the maximum number if eval IDs supported
+	// in a single Delete request. See math above.
+	for i := 0; i < 3; i++ {
+
+		evalList, meta, err := e.client.Evaluations().List(opts)
+		if err != nil {
+			return 1, err
+		}
+		evalsToDelete = append(evalsToDelete, evalList...)
+
+		// Store the next token no matter if it is empty or populated.
+		opts.NextToken = meta.NextToken
+
+		// If there is no next token, ensure we exit and avoid any new loops
+		// which will result in duplicate IDs.
+		if opts.NextToken == "" {
+			break
+		}
+	}
+
+	// The filter flags are operator controlled, therefore ensure we actually
+	// found some evals to delete. Otherwise, inform the operator their flags
+	// are potentially incorrect.
+	if len(evalsToDelete) == 0 {
+		if e.numDeleted > 0 {
+			return 0, nil
+		} else {
+			return 1, errors.New("failed to find any evals that matched filter criteria")
+		}
+	}
+
+	if code, err := e.batchDelete(evalsToDelete); err != nil {
+		return code, err
+	}
+
+	// If there is another page of evaluations matching the filter, call this
+	// function again and delete the next batch of evals. We pause for a 500ms
+	// rather than just run as fast as the code and machine possibly can. This
+	// means deleting 13million evals will take roughly 13-15 mins, which seems
+	// reasonable.
+	if opts.NextToken != "" {
+		time.Sleep(500 * time.Millisecond)
+		return e.handleFlagFilterDelete(opts.NextToken, filter)
```

Comment:
Measuring this will be super interesting! I think the compiler may be able to optimize this into a tail recursion?

Reply:
I wasn't able to find any detail regarding tail recursion optimisation and for the avoidance of unknowns I have altered the behaviour within https://github.com/hashicorp/nomad/pull/13492/commits/4a7cd5f20c305b724ec8f37d09c4c21a0293df11. Although I wouldn't expect any operator to delete 13 millions evals in a single filter query, this command is designed for those types of situations so it's possible.

> we could ensure the evals weren't on the stack

I am a little lost on this point and would appreciate some tips to full understand how this could be achieved or pointers to documentation. Is this the idea of using pointers and such which are known to escape to the heap?

Reply:
(Forgive me if I'm saying anything you already know here, but just a quick brain-dump on the subject.)

Normally locals end up on the stack, which is where we want them for performance. Freeing the stuff that's on the stack is a matter of decrementing the stack pointer and there's no garbage collection required. But sometimes the compiler can't tell if something is safe to have on the stack b/c it might be reused outside the function call (ex. in a caller, on another goroutine, etc), and that's "escape analysis". The entirety of [Go's documentation on this]((https://github.com/golang/go/wiki/CompilerOptimizations#escape-analysis)
) is somewhat laconic:

> Gc compiler does global escape analysis across function and package boundaries. However, there are lots of cases where it gives up. For example, anything assigned to any kind of indirection (*p = ...) is considered escaped. Other things that can inhibit analysis are: function calls, package boundaries, slice literals, subslicing and indexing, etc. Full rules are too complex to describe, so check the -m output.

For recursive functions with potentially large numbers of executions like this, we _want_ large objects to escape to the heap, because that'll reduce the risk we use up the available stack space. And fortunately in Go's case it gives up pretty easily. We can compile with `-gcflags -m` to see whether the compiler escapes any particular line. Running this on Nomad would require digging around a bit in the output. But for a much smaller example, I whipped up this program that queries a dev agent:

<details><summary>silly program</summary>

```go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
)

type NomadAgentHealthResponse struct {
	Client Inner `json:"client"`
	Server Inner `json:"server"`
}
type Inner struct {
	Message string `json:"message"`
	Ok      bool   `json:"ok"`
}

const healthURL = "http://localhost:4646/v1/agent/health"

var client = &http.Client{}

func recurse(i int) error {
	if i < 1 {
		return nil
	}
	raw, err := client.Get(healthURL)
	if err != nil {
		return err
	}
	resp := &NomadAgentHealthResponse{}
	defer raw.Body.Close()

	err = json.NewDecoder(raw.Body).Decode(resp)
	if err != nil {
		return err
	}
	fmt.Println(resp)
	i--
	return recurse(i)
}

func noRecurse(i int) error {
	for i := i; i > 0; i-- {
		raw, err := client.Get(healthURL)
		if err != nil {
			return err
		}
		resp := &NomadAgentHealthResponse{}
		defer raw.Body.Close()

		err = json.NewDecoder(raw.Body).Decode(resp)
		if err != nil {
			return err
		}
		fmt.Println(resp)
	}
	return nil
}

func main() {

	err := recurse(5)
	if err != nil {
		fmt.Println(err)
	}
	err = noRecurse(5)
	if err != nil {
		fmt.Println(err)
	}

}
```

</details>

When we run this, we can see the objects we created do escape to the heap in both the recursive and non-recursive cases. You could use the same tooling (plus a bunch of grepping for the appropriate line 😀 ) to determine if your objects escaped to the heap here as well.

```
$ go run -gcflags -m .
# escape
./main.go:33:23: inlining call to json.NewDecoder
./main.go:37:13: inlining call to fmt.Println
./main.go:51:24: inlining call to json.NewDecoder
./main.go:55:14: inlining call to fmt.Println
./main.go:64:14: inlining call to fmt.Println
./main.go:68:14: inlining call to fmt.Println
./main.go:30:10: &NomadAgentHealthResponse{} escapes to heap
./main.go:33:23: &json.Decoder{...} escapes to heap
./main.go:37:13: ... argument does not escape
./main.go:48:11: &NomadAgentHealthResponse{} escapes to heap
./main.go:51:24: &json.Decoder{...} escapes to heap
./main.go:55:14: ... argument does not escape
./main.go:64:14: ... argument does not escape
./main.go:68:14: ... argument does not escape
&{{ok true} {ok true}}
&{{ok true} {ok true}}
&{{ok true} {ok true}}
&{{ok true} {ok true}}
&{{ok true} {ok true}}
&{{ok true} {ok true}}
&{{ok true} {ok true}}
&{{ok true} {ok true}}
&{{ok true} {ok true}}
&{{ok true} {ok true}}
```

Reply:
Thanks @tgross, of all the documentation i've found online today and previously on this subject this is certainly some of the best conscience detail.

It's worth noting I did compile the command pkg using the `-m`  ldflag earlier and couldn't see any relevant lines to the recursion being escaped (although the output was hard to grok) so I think this change was certainly worth it.